drivers/serial/uart_ns16550.c

```diff
@@ -254,8 +252,12 @@ static void set_baud_rate(struct device *dev, u32_t baud_rate)
 	u8_t lcr_cache;
 
 	if ((baud_rate != 0) && (dev_cfg->sys_clk_freq != 0)) {
-		/* calculate baud rate divisor */
-		divisor = ((dev_cfg->sys_clk_freq / baud_rate) >> 4) + 1;
+		/*
+		 * calculate baud rate divisor. a variant of
+		 * (u32_t)(dev_cfg->sys_clk_freq / (16.0 * baud_rate) + 0.5)
```

Comment:
where is this documented? Is this coming from some datasheet?

Reply:
a long long time when MCU is not so fast,  16 Mhz - 24 Mhz,  the formula is  sclk / (16.0  * baud_rate) + 0.5.  For recent devices, as freq is higher,  it's ok to use (sclk / baud_rate) >> 16.

Iotdk is a special case,  where sclk is just 16 Mhz,  if  16 M / 115200 / 16.0 =  8.68 , there will an error of 0.68 if choose 8,   0.32 if choose 9,  so the right vaule should be 9.  If choose 8, the baurdrate error is out of limit to cause garbage output.